### ISSUE TITLE
Extension method gives odd error message for unrelated type.

### ISSUE THREAD

**Issue by tannergooding** [dotnet/roslyn#33495]
Given the below code, the compiler gives: 
```
error CS1929: 'Vector128<short>' does not contain a definition for 'WithElement' and the best extension method overload 'Vector256.WithElement<int>(Vector256<int>, int, int)' requires a receiver of type 'Vector256<int>'
```

This seems odd given that `Vector256.WithElement` should never be considered, let alone be "the best extension method overload", since the type of `vs` is `Vector128<short>`.

Repro Code ([sharplab](https://sharplab.io/#v2:EYLgHgbALANAJiA1AHwAIGYAEqBMmAqApgM4AuAwgDYCGxxAsAFADeTm72WqAjBNlARKkAsoVIALAPZwAFAEo2HVow6rMANUIBjUpIBO3HAA4APMSl7SAPkwA3YpgC8mOIQBm1AK6VSAbkVqdg7O9gB0AOoAlhIAopSEALaEAHakMgAMMJgAtNxy/iocAL5MJYxMGNi8mAAO1JaR1JTYeJo6+oZGTMqqlTx8bboGxib4NlGx8Umpo1YyEpEOgx0jY3baQ1mRqZjbrmBZ+HZNnoRymADu4oR6hASYIJhkep46mMyYEnqSF5jJ3pRfJgymVKrdqHBJMlKABPWr1UiNZrPV6kDQbFamNYBS7XW73R4onTdHE1PSRWzUUh3cGQ6Fw7ZogD6cE8CQSMIACuSEtEKYQCqCuNU6g0mi10e09DgAKwQEmFdh9arLaVy2aYCbiOKJFKkWbzcSLSVDWUQDW2DF6LY7PaEA73SmUU7nK43O5HQmkF5vD5fH5/AFAkEVLC0qGw+Fi5He1Em/Rm2Y4t34z1PWPElik8mU6mYcP03Y7FlsjncyK8xGWwVAA===)):
```csharp
public class TestClass
{
    public static void TestMethod()
    {
        Vector128<short> vs = default;
        vs = vs.WithElement(0, -1);
    }
}

public static partial class Vector128
{
    public static Vector128<T> WithElement<T>(this Vector128<T> vector, int index, T value) where T : struct { throw null; }
}
public readonly partial struct Vector128<T> 
    where T : struct
{
    private readonly int _dummyPrimitive;
}
public static partial class Vector256
{
    public static Vector256<T> WithElement<T>(this Vector256<T> vector, int index, T value) where T : struct { throw null; }
}
public readonly partial struct Vector256<T>
    where T : struct
{
    private readonly int _dummyPrimitive;
}
```

**Comment by tannergooding** [comment]
CC. @gafter, @jaredpar, @333fred

**Comment by tannergooding** [comment]
If you don't also define `Vector256.WithElement` (see [sharplab](https://sharplab.io/#v2:EYLgHgbALANAJiA1AHwAIGYAEqBMmAqApgM4AuAwgDYCGxxAsAFADeTm72WqAjBNlARKkAsoVIALAPZwAFAEo2HVow6rMANUIBjUpIBO3HAA4APMSl7SAPkwA3YpgC8mOIQBm1AK6VSAbkVqdg7O9gB0AOoAlhIAopSEALaEAHakMgAMMJgAtNxy/iocAL5MJYxMGNi8mAAO1JaR1JTYeJo6+oZGTMqqlTx8bboGxib4NlGx8Umpo1YyEpEOgx0jY3baQ1mRqZjbrmBZ+HZNnoRymADu4oR6hASYIJhkep46mMyYEnqSF5jJ3pRfJgymVKrdqHBJMlKABPWr1UiNZrPV6kDQbFamNYBS7XW73R4onTdHE1PSRWzUUh3cGQ6Fw7ZogD6cE8CQSMIACuSEtEKYQCkUgA==)), you instead get:
```
error CS0411: The type arguments for method 'Vector128.WithElement<T>(Vector128<T>, int, T)' cannot be inferred from the usage. Try specifying the type arguments explicitly.
```

This is also "unfortunate" as it appears the compiler doesn't consider the type of `this` when determining the type of `T` for extension methods.

**Comment by tannergooding** [comment]
The former, definitely seems like a compiler issue; being that the error message is objectively "bad".

The latter is probably a language request, if it is possible to fix as part of "better bestest betterness". If it is thought to be feasible, I might open a separate request on csharplang (assuming there isn't one already).

**Comment by jaredpar** [comment]
@tannergooding 

> The former, definitely seems like a compiler issue; being that the error message is objectively "bad".

Agreed but this is also expected to a degree. Overload resolution is probably the most complex process in the compiler. When there is no match then we have to insert a heuristic to find the "best bad" member. That type of analysis is always going to have edge cases. Improvements to that heuristic are welcome but generally not a priority.

**Comment by otac0n** [comment]
+1 to this issue.  It is going to cause API ergonomics issues for our users when we adopt generic math.

```
using System.Numerics;

R<float> r = default;
r.GetScaled<float>(V<int>.Zero, V<float>.Zero);

struct V<T> where T : INumberBase<T>
{
    public static V<T> Zero = default;
}

struct B<T> where T : INumber<T>
{
}

struct R<T> where T : INumber<T>
{
}

static class B
{
    extension<TSelf>(B<TSelf> b)
        where TSelf : INumber<TSelf>
    {
        public B<TSelf> GetScaled(V<TSelf> scale, V<TSelf> anchor) => default;
    }
}

static class R
{
    extension<TSelf>(R<TSelf> r)
        where TSelf : INumber<TSelf>
    {
        public R<TSelf> GetScaled(V<TSelf> scale, V<TSelf> anchor) => default;
    }
}
```
